toolchain/base/value_ids.h

```diff
@@ -41,20 +50,147 @@ class Real : public Printable<Real> {
   bool is_decimal;
 };
 
-// Corresponds to an integer value represented by an APInt. This is used both
-// for integer literal tokens, which are unsigned and have an unspecified
-// bit-width, and integer values in SemIR, which have a signedness and bit-width
-// matching their type.
-struct IntId : public IdBase, public Printable<IntId> {
+// Corresponds to a canonicalized integer value. This is used both for integer
+// literal tokens, and integer values in SemIR. These always represent the
+// abstract mathematical value -- signed and regardless of the needed precision.
+//
+// Small values are internalized into the ID itself. Large values are
+// represented as an index into an array of `APInt`s with a canonicalized bit
+// width.
+class IntId : public Printable<IntId> {
+ public:
   using ValueType = llvm::APInt;
+
   static const IntId Invalid;
-  using IdBase::IdBase;
+
+  static auto MakeIndexOrInvalid(int index) -> IntId {
+    CARBON_DCHECK(index >= 0 && index <= InvalidIndex);
+    return IntId(ZeroIndexId - index);
+  }
+
+  static auto MakeFromTokenPayload(uint32_t payload) -> IntId {
+    // Token-associated IDs are signed `TokenIdBits` integers, so force the sign
+    // extension from that bit.
+    constexpr int Shift = 32 - TokenIdBits;
+    return IntId(static_cast<int32_t>(payload << Shift) >> Shift);
+  }
+
+  // Tries to make a signed 64-bit integer into an embedded value in the ID, and
+  // if unable to do that returns the `Invalid` ID.
+  static auto TryMakeValue(int64_t value) -> IntId {
+    if (MinValue <= value && value <= MaxValue) {
+      return IntId(value);
+    }
+
+    return Invalid;
+  }
+
+  // Tries to make a signed APInt into an embedded value in the ID, and if
+  // unable to do that returns the `Invalid` ID.
+  static auto TryMakeSignedValue(llvm::APInt value) -> IntId {
```

Comment:
FWIW, since you'd asked organizational comments, it might be worth moving these Make functions to IntStore (if the result is more compact)... for example, I'm having to flip back and forth between files in order to understand how IntStore::AddSigned works, and that might've been something that could be in one spot.